Implements #81: Code-Generator: Add support for custom elements

renelux opened the pull request:
Implementation for feature request #81

mach6 commented on codegen/src/main/java/com/paypal/selion/elements/HtmlSeLionElement.java:
```diff
@@ -25,20 +26,20 @@
-    public static HtmlSeLionElement CONTAINER = new HtmlSeLionElement("Container", true, false);
-    public static HtmlSeLionElement BASE_CLASS = new HtmlSeLionElement("baseClass", false, false);
-    public static HtmlSeLionElement PAGE_TITLE = new HtmlSeLionElement("pageTitle", false, false);
+    public static HtmlSeLionElement TEXT_FIELD = new HtmlSeLionElement("com.paypal.selion.platform.html", "TextField", true, true);
```
Minor suggestion: Make `com.paypal.selion.platform.html` a `public static final String` in this class

renelux replied:
Do you mean in the class AbstractSeLionElement?

```
private String element;
private boolean uiElement;
private String elementPackage;
```

This is how the class looks now so I kept it the same, you want me to update all to public static final?

mach6 commented on codegen/src/main/java/com/paypal/selion/elements/IOSSeLionElement.java:
```diff
@@ -26,40 +27,40 @@
-    public static IOSSeLionElement UIAWebView = new IOSSeLionElement("UIAWebView", true);
-    public static IOSSeLionElement UIAWindow = new IOSSeLionElement("UIAWindow", true);
-    public static IOSSeLionElement BASE_CLASS = new IOSSeLionElement("baseClass", false);
+    public static IOSSeLionElement UIAButton = new IOSSeLionElement("org.uiautomation.ios.UIAModels", "UIAButton", true);
```
Same minor suggestion Re: `public static final String`

mach6 commented:
:+1: @renelux this look good to me but I'm biased since we worked most of the details out together over IRC.

I'm going to ask @sebady @krmahadevan or @fakrudeen78 to chime in.

sebady commented on codegen/src/main/java/com/paypal/selion/elements/HtmlSeLionElement.java:
```diff
@@ -53,6 +56,20 @@
+        
+        temp.add(0, new HtmlSeLionElement(HtmlElementUtils.getPackage(element), HtmlElementUtils.getClass(element), true, false));
+        
+        values = temp.toArray(new HtmlSeLionElement[temp.size()]);
```
So the "overwrite" of an existing standard SeLion element value should at least emit a log message / warning here?  I assume this is not going to be a typical use case here with custom elements? Is there really a need to allow this use case since the order of the elements are matched is important and I can see this tripping up folks who are not careful.

renelux replied:
I believe that the implementation has to be like this. For example in #81 I give an example of a ExtJS SelectList. This element is not compliant with <select> and the SeLion Element doesn't work with it. So I want to create a CustomElement similar to the SelectList if I don't put them in the beginning I could not use a name like ExtJSSelectList as the normal SelectList would be picked up first. So that would limit the users naming convention significantly. 

What about adding a check before elements get add if something with that name already exists? If it does I throw an exception. For the users that want to specifically overwrite, I can add a parameter to disable that check.

sebady replied:
Can we just emit a log message/warning if the custom element will "over-write" a SeLion standard element?  That should be sufficient.

renelux replied:
I have updated my code, in the CodeGeneratorMojo I check if the element is an existing SeLion element. If so it will emit a custom log message to make the user aware of this.

sebady commented:
Rene,  When I merged this I noticed it broke the BasicPageImplTest.  Can you look at what additional change is necessary to get BasicPageImplTest to run successfully?

Commit:
Implements #81: Code-Generator: Add support for custom elements

renelux commented:
@sebady I just updated the BasicPageImpl. And the BasicPageImplTest is now passing.

As the container element is an AbstractElement during initialization it would identify it as a normal element. As the check for AbstractElement came before the check for Container. I changed the order so tt now first checks if the element is a Container, if not it will check if it is a AbstractElement

sebady commented:
merged
